### PR TITLE
[JENKINS-47718] - Deprecate User#getUser(String)

### PR DESCRIPTION
**Pull request by oleg-nenashev**
`User#getUser(String)` method causes lots of confusion for plugin developers. It implicitly creates a user if it cannot be found, and in many cases this is actually a not-expected behavior. This change deprecates the method and to creates a new `User#getOrCreate()` which is more explicit

See [JENKINS-47718](https://issues.jenkins-ci.org/browse/JENKINS-47718).

Changes:

- [x] - Deprecate `User#getUser(String)`
- [x] - Introduce new `User#getOrCreate(String)` as a replacement
- [x] - Describe the behavior and performance concerns in the documentation ([JENKINS-23281](https://issues.jenkins-ci.org/browse/JENKINS-23281))
- [x] - Cleanup usages of `User#getUser(String)` in the production code

### Proposed changelog entries

* BUG: Performance: Do not try to resolve users by full name where it is not required
* RFE, Developer: Deprecate `User#getUser(String)` and improve performance of Jenkins-internal user retrieval logic.
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @jenkinsci/hacktoberfest

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Finally, I am going to kill it. CC @abayer who raised [JENKINS-23281](https://issues.jenkins-ci.org/browse/JENKINS-23281) eons ago

File: core/src/main/java/hudson/model/Cause.java [hunk 1]
```diff
@@ -366,9 +367,15 @@ public UserCause() {
...
 
+        /**
+         * Gets user display name when possible.
+         * @return User display name.
+         *         If the User does not exist, returns its ID.
+         */
         @Exported(visibility=3)
         public String getUserName() {
-        	return User.get(authenticationName).getDisplayName();
+        	final User user = User.get(authenticationName, false, Collections.emptyMap());
```
Comment (jglick) [inline core/src/main/java/hudson/model/Cause.java]: `getById`

File: core/src/main/java/hudson/model/Run.java [hunk 2]
```diff
@@ -1712,7 +1712,10 @@ protected final void execute(@Nonnull RunExecution job) {
                     if (!auth.equals(ACL.SYSTEM)) {
                         String name = auth.getName();
                         if (!auth.equals(Jenkins.ANONYMOUS)) {
-                            name = ModelHyperlinkNote.encodeTo(User.get(name));
+                            final User usr = User.get(name, false, Collections.emptyMap());
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: `getById`

File: core/src/main/java/hudson/model/User.java [hunk 3]
```diff
@@ -494,13 +494,46 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
...
+     *             Otherwise use {@link #getOrCreate(String)} or {@link #get(String, boolean, Map)}.
      */
+    @Deprecated
     public static @Nonnull User get(String idOrFullName) {
-        return get(idOrFullName,true);
+        return getOrCreate(idOrFullName);
     }
 
     /**
+     * Get the user y ID or Full Name.
```
Comment (jglick) [inline core/src/main/java/hudson/model/User.java]: typo

File: core/src/main/java/hudson/model/User.java [hunk 4]
```diff
@@ -494,13 +494,46 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
...
+     * Use {@link #getById} when you know you have an ID.
+     * In this method Jenkins will try to resolve the {@link User} by full name with help of various
+     * {@link hudson.tasks.UserNameResolver}.
+     * This is slow (see JENKINS-23281).
+     *
+     * @param idOrFullName User ID or full name
+     * @return User instance
+     * @since TODO
+     */
+    public static @Nonnull User getOrCreate(@Nonnull String idOrFullName) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/User.java]: If we have to introduce a new method taking `idOrFullName`, we should encode the fact that it is doing a `fullName` lookup in the method name itself. Very few callers should ever need to do this. Really just some SCM plugins.

**Review by jglick — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/model/Cause.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Cause.java]: I am not 100% sure about that. That's why I decided to keep the original behavior.

Comment (jglick) [reply core/src/main/java/hudson/model/Cause.java]: I am fairly confident the original behavior was wrong.

File: core/src/main/java/jenkins/scm/RunWithSCM.java [hunk 5]
```diff
@@ -99,7 +102,8 @@
             public Iterator<User> iterator() {
                 return new AdaptedIterator<String,User>(culpritIds.iterator()) {
                     protected User adapt(String id) {
-                        return User.get(id);
+                        // TODO: Probably it should not auto-create users
```
Comment (jglick) [inline core/src/main/java/jenkins/scm/RunWithSCM.java]: AFAIK the code is correct as it stands. `/asynchPeople` is going to show synthetic users from changelogs, for example.

File: core/src/main/java/hudson/model/User.java [hunk 6]
```diff
@@ -536,12 +540,45 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
...
+     * In this method Jenkins will try to resolve the {@link User} by full name with help of various
+     * {@link hudson.tasks.UserNameResolver}.
+     * This is slow (see JENKINS-23281).
+     *
+     * @deprecated This method is deprecated, because it causes unexpected {@link User} creation
+     *             by API usage code and causes performance degradation of used to retrieve users by ID.
+     *             Use {@link #getById} when you know you have an ID.
+     *             Otherwise use {@link #getOrCreateByIdOrFullName(String)} or {@link #get(String, boolean, Map)}.
      */
+    @Deprecated
```
Comment (jglick) [inline core/src/main/java/hudson/model/User.java]: What about deprecating `get(String, boolean, Map)` also? Or do you think that overload is weird enough that people would not even consider calling it until they had read the Javadoc and noticed the “or full name” part?

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/model/User.java (hunk 6 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/User.java]: Well, I definitely have plugins where I need to refactor that (ownership & Co). Maybe it also needs to be deprecated, but right now there is no full replacement assuming somebody really uses `context`

**Comment by oleg-nenashev** [comment]
@reviewbybees done

File: core/src/main/java/hudson/model/Cause.java [hunk 7]
```diff
@@ -409,12 +416,15 @@ public String getUserId() {
 
         @Exported(visibility = 3)
         public String getUserName() {
-            return userId == null ? "anonymous" : User.get(userId).getDisplayName();
+            final User user = userId == null ? null : User.getById(userId, false);
+            return user == null ? "anonymous" : user.getDisplayName();
```
Comment (Wadeck) [inline core/src/main/java/hudson/model/Cause.java]: Perhaps add a `@Nonnull` annotation as we either have "anonymous" or the return a of a nonnull method (User.getDisplayName())

**Review by Wadeck — APPROVED** [review]
:bee:

File: core/src/main/java/hudson/model/Cause.java (hunk 7 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Cause.java]: Subject for polishing later. Ideally I want to define nonnull requirement on the package level

**Pull request merged**